File: Programmers/Swift/괄호 회전하기.swift
import Foundation

var res = 0
var str = [Character]()
var st = [Character]()

func judge(_ s:[Character]) {
    var chk = true
    for c in s {
        if c=="{" || c=="[" || c=="(" {st.append(c)}
        else {
            if st.count==0
            || c=="}" && st[st.count-1] != "{"
            || c==")" && st[st.count-1] != "("
            || c=="]" && st[st.count-1] != "[" {
                chk = false
                break
            }
            st.removeLast()
        }
    }
    if st.count>0 {chk=false}
    if chk == true {res+=1}
}

func solution(_ s:String) -> Int {
    
    str = Array(s)
    for i in 0..<s.count {
        var arg = [Character]()
        for j in i..<s.count {
            arg.append(str[j])
        }
        for j in 0..<i {
            arg.append(str[j])
        }
        judge(arg)
    }
    return res
}

File: Programmers/C++/괄호 회전하기.cpp
#include <iostream>
#include <string>
#include <vector>
#include <stack>
#include <algorithm>
#define FOR(i,n) for(int i=0;i<(n);++i)
#define FORS(i,s,n) for(int i=(s);i<(n);++i)

using namespace std;

int res;

void judge(string k) {
    stack<char> st;
    
    bool chk = 1;
    for(auto c: k) {
        if(c=='(' || c=='[' || c=='{') st.push(c);
        else {
            if(st.empty()
               || (c==']'&&st.top()!='[')
               || (c=='}'&&st.top()!='{')
               || (c==')'&&st.top()!='(')) {
                chk=0;
                break;
            }
            st.pop();
        }
    }
    if(st.size()) chk=0;
    if(chk) ++res;
}

int solution(string s) {
    
    FOR(i,s.size()) {
        string k = s.substr(i,s.size()-i);
        FOR(j,i) k+=s[j];
        judge(k);
    }
    
    return res;
}
